internal/provider/providers/desec/provider.go

```diff
@@ -0,0 +1,143 @@
+	token         string
+	useProviderIP bool
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *Provider, err error) {
+	extraSettings := struct {
+		Token         string `json:"token"`
+		UseProviderIP bool   `json:"provider_ip"`
+	}{}
+	err = json.Unmarshal(data, &extraSettings)
+	if err != nil {
+		return nil, err
+	}
+	p = &Provider{
+		domain:        domain,
+		host:          host,
+		ipVersion:     ipVersion,
+		token:         extraSettings.Token,
+		useProviderIP: extraSettings.UseProviderIP,
+	}
+	err = p.isValid()
+	if err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *Provider) isValid() error {
+	switch {
+	case p.token == "":
+		return fmt.Errorf("%w", errors.ErrTokenNotSet)
+	case p.host == "*":
+		return fmt.Errorf("%w", errors.ErrHostWildcard)
+	}
+	return nil
+}
+
+func (p *Provider) String() string {
+	return fmt.Sprintf("[domain: %s | host: %s | provider: deSEC]", p.domain, p.host)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Host() string {
+	return p.host
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.host, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName()),
+		Host:      p.Host(),
+		Provider:  "<a href=\"https://desec.io/\">deSEC</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+	u := url.URL{
+		Scheme: "https",
+		User:   url.UserPassword(p.BuildDomainName(), p.token),
+		Host:   "update.dedyn.io",
+		Path:   "/nic/update",
+	}
+	values := url.Values{}
+	values.Set("hostname", utils.BuildURLQueryHostname(p.host, p.domain))
+	if !p.useProviderIP {
+		values.Set("myip", ip.String())
+	}
+	u.RawQuery = values.Encode()
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("creating http request: %w", err)
+	}
+	headers.SetUserAgent(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return netip.Addr{}, err
+	}
+	defer response.Body.Close()
+
+	b, err := io.ReadAll(response.Body)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("reading response body: %w", err)
+	}
+	s := string(b)
+
+	if response.StatusCode != http.StatusOK {
+		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.ToSingleLine(s))
+	}
+
+	switch {
+	case strings.HasPrefix(s, constants.Notfqdn):
+		return netip.Addr{}, fmt.Errorf("%w", errors.ErrHostnameNotExists)
+	case strings.HasPrefix(s, "badrequest"):
+		return netip.Addr{}, fmt.Errorf("%w", errors.ErrBadRequest)
```

Comment:
Can you check if you send a malformed request, for example `myip=abc`, what's the status code returned (you can do that in your browser or using curl)? Is it 200 or could it be 400? Asking this since when unauthorized it returns the 401 status code... 🤔

Reply:
I tried a malformed host ("hostname"=abc) and got a 404: nohost response. I also tried a malformed ip ("myip"=abc), and the server ignored my error, and correctly set the ip based on its own check ip function. I have added the new 404 to the status code check.
One last question before committing the patch. The string based error handler seems to be redundant, since the errors are handled through status codes. Should I remove the string based check from the commit?

Reply:
If their documentation mentions they use the dyndns protocol, I think we should keep it to match the dyndns protocol 🤔 Well technically we should support all codes as described in https://help.dyn.com/remote-access-api/return-codes/ but I am planning on refactoring all providers using the dyndns protocol and have a common single implementation for them, so don't bother I would say 😉

Reply:
Sounds good, committed the patch with error handling.